sdk/typespec/typespec_client_core/src/tracing/mod.rs

```diff
@@ -0,0 +1,170 @@
+///   entrypoint for distributed tracing in the SDK.
+/// - Tracer: This trait is responsible for creating spans and managing the active span.
+/// - Span: This trait represents a single unit of work in the distributed tracing system.
+///
+pub mod attributes;
+pub mod with_context;
+pub use with_context::{FutureExt, WithContext};
+
+/// The TracerProvider trait is the entrypoint for distributed tracing in the SDK.
+///
+/// It provides a method to get a tracer for a specific name and package version.
+pub trait TracerProvider {
+    /// Returns a tracer for the given name.
+    ///
+    /// Arguments:
+    /// - `package_name`: The name of the package for which the tracer is requested.
+    /// - `package_version`: The version of the package for which the tracer is requested.
+    fn get_tracer(
+        &self,
+        package_name: &'static str,
+        package_version: &'static str,
+    ) -> Box<dyn Tracer + Send + Sync>;
+}
+
+pub trait Tracer {
+    /// Starts a new span with the given name and type.
+    ///
+    /// # Arguments
+    /// - `name`: The name of the span to start.
+    /// - `kind`: The type of the span to start.
+    ///
+    /// # Returns
+    /// An `Arc<dyn Span + Send + Sync>` representing the started span.
+    ///
+    fn start_span(&self, name: &'static str, kind: SpanKind) -> Arc<dyn Span + Send + Sync>;
+
+    /// Starts a new span with the given type, using the current span as the parent span.
+    ///
+    /// # Arguments
+    /// - `name`: The name of the span to start.
+    /// - `kind`: The type of the span to start.
+    ///
+    /// # Returns
+    /// An `Arc<dyn Span + Send + Sync>` representing the started span.
+    ///
+    fn start_span_with_current(
+        &self,
+        name: &'static str,
+        kind: SpanKind,
+    ) -> Arc<dyn Span + Send + Sync>;
+
+    /// Starts a new child with the given name, type, and parent span.
+    ///
+    /// # Arguments
+    /// - `name`: The name of the span to start.
+    /// - `kind`: The type of the span to start.
+    /// - `parent`: The parent span to use for the new span.
+    ///
+    /// # Returns
+    /// An `Arc<dyn Span + Send + Sync>` representing the started span
+    ///
+    fn start_span_with_parent(
+        &self,
+        name: &'static str,
+        kind: SpanKind,
+        parent: Arc<dyn Span + Send + Sync>,
+    ) -> Arc<dyn Span + Send + Sync>;
+}
+pub enum SpanStatus {
+    Unset,
+    Ok,
+    Error { description: String },
+}
+
+#[derive(Debug, Default)]
+pub enum SpanKind {
+    #[default]
+    Internal,
+    Client,
+    Server,
+    Producer,
+    Consumer,
+}
+
+pub trait SpanGuard {
+    /// Ends the span when dropped.
+    fn end(self) -> crate::Result<()>;
+}
+
+pub trait Span: AsAny {
+    /// The 8 byte value which identifies the span.
+    fn span_id(&self) -> [u8; 8];
+
+    /// Ends the current span.
+    fn end(&self) -> crate::Result<()>;
+
+    /// Adds an event to the current span.
```

Comment:
span_events are going away from OTel, so maybe its best to *not* add them here...

Reply:
Already removed.